snuba/datasets/errors_replacer.py

```diff
@@ -206,93 +213,124 @@ def set_project_needs_final(
-) -> Tuple[bool, Sequence[int], Set[str]]:
-    """\
+) -> ProjectsQueryFlags:
+    """
     1. Fetch `needs_final` for each Project
     2. Fetch groups to exclude for each Project
-    3. Trim groups to exclude ZSET for each Project
+    3. Clean up the old replacement types and groups
     4. Fetch replacement types for each Project
+    5. Fetch latest exclude groups replacement type for each Project
 
-    Returns (needs_final, group_ids_to_exclude, replacement_types)
+    Returns (
+        needs_final,
+        group_ids_to_exclude,
+        replacement_types,
+        latest_replacement_time
+    )
     """
 
     s_project_ids = set(project_ids)
     now = time.time()
     p = redis_client.pipeline()
 
     needs_final_keys_and_type_keys = [
-        get_project_needs_final_key_and_type_key(project_id, state_name)
+        build_project_needs_final_key_and_type_key(project_id, state_name)
         for project_id in s_project_ids
     ]
 
     for needs_final_key, _ in needs_final_keys_and_type_keys:
         p.get(needs_final_key)
 
     exclude_groups_keys_and_types = [
-        get_project_exclude_groups_key_and_type_key(project_id, state_name)
+        build_project_exclude_groups_key_and_type_key(project_id, state_name)
         for project_id in s_project_ids
     ]
-
     for exclude_groups_key, _ in exclude_groups_keys_and_types:
+        # remove stale excluded groups
         p.zremrangebyscore(
             exclude_groups_key, float("-inf"), now - settings.REPLACER_KEY_TTL
         )
+        # TODO: put this into another for loop to make post processing easier
+        # get up to date excluded groups
         p.zrevrangebyscore(
             exclude_groups_key, float("inf"), now - settings.REPLACER_KEY_TTL
         )
 
+    # get the replacement types for metrics purposes
     for _, needs_final_type_key in needs_final_keys_and_type_keys:
         p.get(needs_final_type_key)
 
     for _, type_key in exclude_groups_keys_and_types:
+        # cleanup the old exclude group replacement types
         p.zremrangebyscore(type_key, float("-inf"), now - settings.REPLACER_KEY_TTL)
+        # get the new exclude group replacement types
+        # TODO: put this into another for loop to make post processing easier
         p.zrevrangebyscore(type_key, float("inf"), now - settings.REPLACER_KEY_TTL)
 
+    # retrieve the latest replaced group id's timestamp such that queries
+    # which are processing data after it, do not have to be marked as final
+    for exclude_groups_key, _ in exclude_groups_keys_and_types:
+        p.zrevrange(
+            exclude_groups_key, 0, 0, withscores=True,
+        )
+
     results = p.execute()
 
-    return process_exclude_groups_and_replacement_types_results(
+    return _process_exclude_groups_and_replacement_types_results(
         results, len(s_project_ids)
     )
 
 
-def process_exclude_groups_and_replacement_types_results(
+def _process_exclude_groups_and_replacement_types_results(
```

Comment:
This method and the following seem to have the only goal of producing the the `ProjectsQueryFlags` instance from the bare results you got from Redis. 

Now the `ProjectQueryFlags` data class is just a bag of data. It seems these methods should be part of that class as a static factory method. 

```
class ProjectsQueryFlags:
....
@classmethod
def load_from_redis(cls, project_ids: Sequence[int], state_name: Optional[ReplacerState]):  ProjectsQueryFlags:
   ...the whole logic ...
```

This encapsulates these methods into a class that is the only place concerned with them.

Please do that in a separate PR

Reply:
#2155